python/plugins/processing/algs/qgis/FieldsMapper.py

```diff
@@ -28,159 +27,141 @@
 
 __revision__ = '$Format:%H$'
 
-from qgis.core import (QgsField,
-                       QgsFields,
-                       QgsExpression,
-                       QgsDistanceArea,
-                       QgsFeatureSink,
-                       QgsProject,
-                       QgsFeature,
-                       QgsApplication,
-                       QgsProcessingUtils)
-from processing.algs.qgis.QgisAlgorithm import QgisAlgorithm
-from processing.core.GeoAlgorithmExecutionException import GeoAlgorithmExecutionException
-from processing.core.parameters import ParameterTable
-from processing.core.parameters import Parameter
-from processing.core.outputs import OutputVector
+from qgis.core import (
+    QgsApplication,
+    QgsDistanceArea,
+    QgsExpression,
+    QgsFeature,
+    QgsFeatureSink,
+    QgsField,
+    QgsFields,
+    QgsProcessingException,
+    QgsProcessingParameterDefinition,
+    QgsProcessingUtils,
+    QgsProject,
+)
 
+from processing.algs.qgis.QgisAlgorithm import QgisFeatureBasedAlgorithm
 
-class FieldsMapper(QgisAlgorithm):
+
+class FieldsMapper(QgisFeatureBasedAlgorithm):
 
     INPUT_LAYER = 'INPUT_LAYER'
     FIELDS_MAPPING = 'FIELDS_MAPPING'
     OUTPUT_LAYER = 'OUTPUT_LAYER'
 
-    def __init__(self):
-        GeoAlgorithm.__init__(self)
-        self.mapping = None
-
     def group(self):
         return self.tr('Vector table tools')
 
-    def __init__(self):
-        super().__init__()
-
-    def initAlgorithm(self, config=None):
-        self.addParameter(ParameterTable(self.INPUT_LAYER,
-                                         self.tr('Input layer'),
-                                         False))
-
-        class ParameterFieldsMapping(Parameter):
+    def initParameters(self, config=None):
 
-            default_metadata = {
-                'widget_wrapper': 'processing.algs.qgis.ui.FieldsMappingPanel.FieldsMappingWidgetWrapper'
-            }
+        class ParameterFieldsMapping(QgsProcessingParameterDefinition):
 
-            def __init__(self, name='', description='', parent=None):
-                Parameter.__init__(self, name, description)
-                self.parent = parent
-                self.value = []
+            def __init__(self, name, description, parentLayerParameterName='INPUT'):
+                super().__init__(name, description)
+                self._parentLayerParameter = parentLayerParameterName
 
-            def getValueAsCommandLineParameter(self):
-                return '"' + str(self.value) + '"'
+            def type(self):
+                return 'fields_mapping'
 
-            def setValue(self, value):
-                if value is None:
+            def checkValueIsAcceptable(self, value, context):
```

Comment:
This is missing the default value for context - it should be `context=None` as the context is not always available or passed to this method (e.g. currently this crashes when running through the toolbox)